From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Moderocky <admin@moderocky.com>
Date: Thu, 24 Sep 2020 13:24:13 +0100
Subject: [PATCH] Markers for fake entities.


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 34d9db7fd60986340d40bb51f1d54584b1332aa9..b458d696073948880229d97db35fc996cfa5cc90 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -184,6 +184,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     private EntityDamageEvent lastDamageEvent;
     private final CraftPersistentDataContainer persistentDataContainer = new CraftPersistentDataContainer(DATA_TYPE_REGISTRY);
     public final AttachmentHolder attachments; // Kenzie
+    public boolean fake; // Kenzie - public so implementations can change it
 
     public CraftEntity(final CraftServer server, final Entity entity) {
         this.server = server;
@@ -192,10 +193,21 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         this.attachments = new AttachmentHolder(this);
         server.attachmentFactory.addAttachments(this, attachments);
         server.attachmentFactory.registerAttachmentHolder(attachments);
+        fake = false;
         // Kenzie end
     }
 
-    // Kenzie start - NBT holder
+    // Kenzie start - NBT holder and fakes
+
+    @Override
+    public boolean isFake() {
+        return fake;
+    }
+
+    public void setFake(boolean boo) {
+        fake = boo;
+    }
+
     @Override
     public NBTCompound getNBT() {
         NBTTagCompound compound = new NBTTagCompound();
