From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sat, 3 Oct 2020 21:39:07 -0500
Subject: [PATCH] Entity#isTicking


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 1e02242b47f38c90d925b3084a7d284696234475..20fb0cb358ba71e7ec9535851ba6dbfa78e7a449 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -693,5 +693,10 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      * Check if entity is in lava
      */
     public boolean isInLava();
+
+    /**
+     * Check if entity is inside a ticking chunk
+     */
+    public boolean isTicking();
     // Paper end
 }
