From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Moderocky <admin@moderocky.com>
Date: Thu, 8 Oct 2020 20:14:38 +0100
Subject: [PATCH] Missed annotation.


diff --git a/src/main/java/mx/kenzie/wellspring/nbt/BlockNBTStorage.java b/src/main/java/mx/kenzie/wellspring/nbt/BlockNBTStorage.java
index 870ea93917c6cfbb12f57da8772c7342d8dbae38..d07a30fee7fc280742cadba2351b2c864bcacca1 100644
--- a/src/main/java/mx/kenzie/wellspring/nbt/BlockNBTStorage.java
+++ b/src/main/java/mx/kenzie/wellspring/nbt/BlockNBTStorage.java
@@ -13,7 +13,7 @@ import org.jetbrains.annotations.NotNull;
  */
 public interface BlockNBTStorage {
 
-    default @NotNull NBTCompound getNBT(Block block) {
+    default @NotNull NBTCompound getNBT(@NotNull Block block) {
         return getNBT(new Vector(block.getX(), block.getY(), block.getZ()));
     }
 
