From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Moderocky <admin@moderocky.com>
Date: Sat, 24 Oct 2020 14:32:39 +0100
Subject: [PATCH] Ray trace to target position.


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index e3afa5e5975d6e263f3363298fd6185bd0a24821..526bc6e7f35c39d39f24a3dba36988e3a6017130 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -806,6 +806,9 @@ public interface World extends PluginMessageRecipient, Metadatable, Attachable,
     @Nullable
     public RayTraceResult rayTraceBlocks(@NotNull Location start, @NotNull Vector direction, double maxDistance, @NotNull FluidCollisionMode fluidCollisionMode);
 
+    @Nullable
+    public RayTraceResult rayTraceBlocks(@NotNull Location start, @NotNull Location end, @NotNull FluidCollisionMode fluidCollisionMode, boolean ignorePassableBlocks); // Kenzie
+
     /**
      * Performs a ray trace that checks for block collisions using the blocks'
      * precise collision shapes.
