From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Moderocky <admin@moderocky.com>
Date: Thu, 29 Oct 2020 10:04:11 +0000
Subject: [PATCH] Fix git error.


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 6fda50e92435efcf015dd217eb6dc44b7dd9db01..52807e52a6f4e58a1c00d57ba6e00bb844c8dc51 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1794,46 +1794,6 @@ public final class Bukkit {
     }
     // Paper end
 
-    @NotNull
-    public static org.bukkit.command.CommandMap getCommandMap() {
-        return server.getCommandMap();
-    }
-
-    /**
-     * Reload the Permissions in permissions.yml
-     */
-    public static void reloadPermissions() {
-        server.reloadPermissions();
-    }
-
-    /**
-     * Reload the Command Aliases in commands.yml
-     *
-     * @return Whether the reload was successful
-     */
-    public static boolean reloadCommandAliases() {
-        return server.reloadCommandAliases();
-    }
-
-    /**
-     * Checks if player names should be suggested when a command returns {@code null} as
-     * their tab completion result.
-     *
-     * @return true if player names should be suggested
-     */
-    public static boolean suggestPlayerNamesWhenNullTabCompletions() {
-        return server.suggestPlayerNamesWhenNullTabCompletions();
-    }
-
-    /**
-     *
-     * @return the default no permission message used on the server
-     */
-    @NotNull
-    public static String getPermissionMessage() {
-        return server.getPermissionMessage();
-    }
-
     // Kenzie start - attachments
 
     /**
@@ -1950,64 +1910,6 @@ public final class Bukkit {
     }
     // Kenzie end
 
-    /**
-     * Creates a PlayerProfile for the specified uuid, with name as null
-     * @param uuid UUID to create profile for
-     * @return A PlayerProfile object
-     */
-    @NotNull
-    public static com.destroystokyo.paper.profile.PlayerProfile createProfile(@NotNull UUID uuid) {
-        return server.createProfile(uuid);
-    }
-
-    /**
-     * Creates a PlayerProfile for the specified name, with UUID as null
-     * @param name Name to create profile for
-     * @return A PlayerProfile object
-     */
-    @NotNull
-    public static com.destroystokyo.paper.profile.PlayerProfile createProfile(@NotNull String name) {
-        return server.createProfile(name);
-    }
-
-    /**
-     * Creates a PlayerProfile for the specified name/uuid
-     *
-     * Both UUID and Name can not be null at same time. One must be supplied.
-     *
-     * @param uuid UUID to create profile for
-     * @param name Name to create profile for
-     * @return A PlayerProfile object
-     */
-    @NotNull
-    public static com.destroystokyo.paper.profile.PlayerProfile createProfile(@Nullable UUID uuid, @Nullable String name) {
-        return server.createProfile(uuid, name);
-    }
-
-    public static int getCurrentTick() {
-        return server.getCurrentTick();
-    }
-
-    /**
-     * Checks if the server is in the process of being shutdown.
-     *
-     * @return true if server is in the process of being shutdown
-     */
-    public static boolean isStopping() {
-        return server.isStopping();
-    }
-
-    /**
-     * Returns the {@link com.destroystokyo.paper.entity.ai.MobGoals} manager
-     *
-     * @return the mob goals manager
-     */
-    @NotNull
-    public static com.destroystokyo.paper.entity.ai.MobGoals getMobGoals() {
-        return server.getMobGoals();
-    }
-    // Paper end
-
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
